crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -973,21 +1049,35 @@ impl<InstalledPackages: InstalledPackagesProvider> ResolverState<InstalledPackag
                     self.visit_package(dep_package, request_sink)?;
                 }
 
+                // If a package has a marker, add a dependency from it to
+                // the same package without markers.
```

Comment:
To ensure they resolve to the same version, or something?

Reply:
I can't quite explain this yet. If I don't do this (and adjust some of the code in `resolution/graph.rs` to not require `marker: None`), then I get a few test failures. One of them is this:

```
Snapshot: compile_constraints_markers
Source: crates/uv/tests/pip_compile.rs:322
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: snapshot
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    5     5 │ anyio==4.3.0
    6     6 │     # via -r requirements.in
    7     7 │ idna==3.6
    8     8 │     # via anyio
    9       │-sniffio==1.3.0
          9 │+sniffio==1.3.1
   10    10 │     # via
   11    11 │     #   -c constraints.txt
   12    12 │     #   anyio
   13    13 │
   14    14 │ ----- stderr -----
   15       │-Resolved 3 packages in [TIME]
         15 │+Resolved 4 packages in [TIME]
```

And there are 3 more failures that look like this:

```
Snapshot: install_executable
Source: crates/uv/tests/pip_install.rs:1802
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: snapshot
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    1     1 │ exit_code: 0
    2     2 │ ----- stdout -----
    3     3 │
    4     4 │ ----- stderr -----
    5       │-Resolved 7 packages in [TIME]
          5 │+Resolved 8 packages in [TIME]
    6     6 │ Downloaded 7 packages in [TIME]
    7     7 │ Installed 7 packages in [TIME]
    8     8 │  + astroid==3.0.3
    9     9 │  + dill==0.3.8
```

I think I originally added this because I was trying to follow in the footsteps of how extras were handled. But markers aren't exactly like extras, and I'm not sure this sort of dependency makes sense.

Looking at the `compile_constraints_markers` test more closely, I see that it looks like there is probably something relevant here:

```
    // If constraints are ignored, these will conflict
    constraints_txt.write_str("sniffio==1.2.0;python_version<='3.7'")?;
    constraints_txt.write_str("sniffio==1.3.0;python_version>'3.7'")?;
```

So if there is no `marker: None` "dependency on itself" being added (as is done here in this PR), then I believe the above two constraints will end up being two _distinct_ `PubGrubPackage` values due to the markers being distinct. Which in turn means the pubgrub constraints won't unify and I think leads to a different resolution? Although I can't fully explain why `1.3.1` is selected. With the dependency on itself, I think that lets pubgrub unify the constraints, which I think _is_ similar to the same kind of effect we achieve with extras?

I'm like 80% confident here. I've updated the comment here to hopefully expose some of that uncertainty:

```rust
                // If a package has a marker, add a dependency from it to the
                // same package without markers.
                //
                // At time of writing, AG doesn't fully understand why we need
                // this, but one explanation is that without it, there is no
                // way to connect two different `PubGrubPackage` values with
                // the same package name but different markers. With different
                // markers, they would be considered wholly distinct packages.
                // But this dependency-on-itself-without-markers forces PubGrub
                // to unify the constraints across what would otherwise be two
                // distinct packages.
```